src/GafferArnold/IECoreArnoldPreview/Renderer.cpp

```diff
@@ -2122,6 +2122,18 @@ class ArnoldLight : public ArnoldObjectBase
 
 			if( m_attributes->lightShader() )
 			{
+				if( m_lightShader )
```

Comment:
Rather than add extra control flow, can't we put this in the block that already exists, at L2147 (old code, L2159 now)? I wonder if we should also perform some additional checks to avoid the full purge unless it's absolutely necessary. Is it only changes to the `color` parameter that affect the importance tree, so that we could achieve smoother updates for exposure changes etc? I guess it's made slightly trickier by needing to check for upstream changes in connected input shaders? At the very least, it would be trivial to check that it's the light shader that has changed, rather than some other parameter.

Reply:
I've moved inside the existing block.  Sorry, this went through a bunch of iterations.

> At the very least, it would be trivial to check that it's the light shader that has changed, rather than some other parameter.

Maybe I'm just brain dead after hopping between 6 different issues today, but I'm not spotting how this is trivial.  This code is being run in the case that there is both a stored light shader, and an incoming light shader.  The only way to check if the light shader has changed would to check if all parameters of all shaders in the network currently match their Arnold values?  Is there some function which does this that I'm not aware of?

Reply:
I was thinking `oldAttributes->lightShader()->isEqualTo( attributes->lightShader() )`? But reviewing the work involved in deleting and recreating the light (updating all the light links), I'm fairly inclined towards a proper check that does a targeted comparison of the ShaderNetwork upstream of the `color` parameter. Let's cover in our chat later...